lib/classes/designator.rb

```diff
@@ -0,0 +1,54 @@
+class Designator
+  attr_accessor :package, :params, :order_id, :orders_package
+
+  def initialize(package, package_params)
+    @package = package
+    @params = package_params
+    @order_id = package_params[:order_id].to_i
+    @orders_package = @package.orders_packages.new
+    @existing_designation = OrdersPackage.find_by_id(@params[:orders_package_id])
+  end
+
+  def designate
+    if designating_to_existing_designation?
+      add_error_and_return_existing_designation
+    else
+      form_nested_params_for_undesignate if @package.quantity.zero?
+      designate_item
+    end
+  end
+
+  # undesignate_package params is passed from 'form_nested_params_for_undesignate'
+  def undesignate(undesignate_package = nil)
+    packages = undesignate_package ? undesignate_package : @params
+    OrdersPackage.undesignate_partially_designated_item(packages)
+    @package.undesignate_from_stockit_order
+  end
+
+  def form_nested_params_for_undesignate
+    undesignate_package = {}
+    @params[:quantity] = @package.received_quantity
+    undesignate_package["0"] = @params
+    undesignate(undesignate_package)
+  end
+
+  private
+
+  def add_error_and_return_existing_designation
+    @existing_designation.errors.add("package_id", "Already designated to this Order")
+    @existing_designation
+  end
+
+  def designating_to_existing_designation?
+    @existing_designation&.order_id == @params[:order_id].to_i
+  end
+
+  def designate_item
+    @orders_package.order_id = @order_id
+    @orders_package.quantity = @params[:quantity_to_designate] ? @params[:quantity_to_designate] : @params[:quantity]
```

Comment:
whats difference in params[:quantity_to_designate] and params[:quantity] ?

We should always get only on param in this case. I don't think we need to send 2 different params from ember just for same action which is designate.

Not a blocker but will be good to change.

Reply:
`quantity_to_designate` will come in params if package is designated else it will not be present in params and in that case `quantity` will value be assigned.

Reply:
But it's not good to send 2 different params for 2 scenario's for same designate action.

Reply:
Adding change of only using `quantity_to_designate`.